### PR TITLE
Extract and use report lengths contained in HID report descriptor

### PR DESCRIPTION
**Pull request by FabianHenneke**
This PR is meant to address #165 by replacing the hardcoded HID report length of 64 bytes by the lengths reported in the FIDO devices HID report descriptor.

It is broken up into an initial commit that replaces the hardcoded HID report length with configurable input and output report lengths, while still using the default report length in all backends. This is meant to simplify the review process. This commit also adapts the regression tests and fuzzers accordingly.

The later commits implement the extraction of the report lengths from the report descriptor for the various backends. Here, the status of the individual commits is as follows:

- [x]  Windows: Builds and performs as expected as confirmed by a manual test with both a 64-byte and a 62-byte report FIDO device.
- [x] Linux: Builds and performs as expected as confirmed by a manual test with both a 64-byte and a 62-byte report FIDO device.
- [ ] macOS: Does build in the CI, but I haven't been able to test it since I do not have access to a physical device running macOS.
- [ ] hidapi: Is fundamentally incompatible with devices with unknown HID report sizes since it does not extract the report lengths from the report descriptor on all platforms. The respective commit adds a note about this, but this may warrant mentioning in the README (or elsewhere).
- [ ] OpenBSD: Not implemented yet. I lack experience with OpenBSD and do not know which library I should use to fetch the HID report descriptor or report lengths for both USB and Bluetooth HID devices.

### CODE REVIEW

**Comment by martelletto** [comment]
Thank you! Much appreciated. I am planning to do a formal review over the weekend, but will try my best to provide feedback of some kind before then. Out of curiosity, are the changes enough for WearAuthn to work on Linux (as I understand it, hidraw(4) abstracts both USB and Bluetooth HID) for SSH auth?

**Comment by FabianHenneke** [comment]
> Thank you! Much appreciated. I am planning to do a formal review over the weekend, but will try my best to provide feedback of some kind before then. 

Thank you very much!

> Out of curiosity, are the changes enough for WearAuthn to work on Linux (as I understand it, hidraw(4) abstracts both USB and Bluetooth HID) for SSH auth?

Thanks for your interest in WearAuthn. The current PR makes ssh-keygen work with WearAuthn. In order for actual SSH authentication to work, #169 is needed to get libfido2 to recognize WearAuthn as a HID device and report it back to SSH. 

I will see whether I can get OpenSSH to support the new pinUvAuthToken once it is no longer hidden behind a flag in libfido2. Then, when credential management has made it to WearAuthn, it should also support the resident key mode of ssh-keygen.

File: src/fido/types.h [hunk 1]
```diff
@@ -16,12 +16,18 @@ extern "C" {
...
+	uint16_t report_in_len;
+	uint16_t report_out_len;
+} fido_dev_io_info_t;
+
+typedef struct fido_dev_io_info *fido_dev_io_open_t(const char *);
+typedef void   fido_dev_io_close_t(struct fido_dev_io_info *);
+typedef int    fido_dev_io_read_t(struct fido_dev_io_info *, unsigned char *, size_t, int);
+typedef int    fido_dev_io_write_t(struct fido_dev_io_info *, const unsigned char *, size_t);
+typedef int    fido_dev_rx_t(struct fido_dev *, uint8_t, unsigned char *, size_t, int);
+typedef int    fido_dev_tx_t(struct fido_dev *, uint8_t, const unsigned char *, size_t);
```
Comment (martelletto) [inline src/fido/types.h:30]: These definitions are part of the public API. Would it be possible to keep them as they are, and cast `void *` to `fido_dev_io_info_t` internally, similar to what hid_openbsd.c and hid_osx.c do?

Comment (FabianHenneke) [reply src/fido/types.h:30]: Thanks for catching that, I wasn't aware that these functions were part of the public API until now. Yes, that should be possible, I will force-push a new version once I have it ready.

Comment (FabianHenneke) [reply src/fido/types.h]: I have amended the first commit to preserve the public API. This is done by including a flag in `fido_dev_t` that allows switching between the native, report size aware io function signature expecting a `fido_dev_io_info_t` and the legacy io functions expecting just a void pointer. If desired, another commit could add a version of `fido_dev_set_io_functions` that is report size aware.

**Comment by FabianHenneke** [comment]
All CI checks pass and I will not make any more changes before the review.

**Comment by martelletto** [comment]
Thanks! I will have a look.

**Comment by martelletto** [comment]
Hi Fabian,

If possible, I would prefer to merge these commits separately. Doing so would allow modifications to be merged and tested more quickly. As such, while I did look at all the commits, I focused my attention on the first one:

- I like the `legacy_io` approach. I suppose we could rename it `fixed_rpt_size` (for clarity), or derive it from `io.open != fido_hid_open` (for obscurity ;-));
- Perhaps we could define a MIN counterpart of `MAX_CTAP_REPORT_LEN` (`CTAP_INIT_HEADER_LEN + 1`?), and call them `CTAP_{MIN,MAX}_REPORT_LEN` (for consistency with other CTAP_* definitions);
- We can probably get rid of the open/close/read/write wrappers by moving `report_{in,out}_len` inside `fido_dev_t` and defining` fido_hid_report_{in,out}_len()` in our HID backends; see 9753f45ae (untested and incomplete code; just a sketch). In this approach, dev.c and hid_*.c are kept separate and do not share types;
- Would it make sense to define `report_{in,out}_len` in `fido_dev_t` as `uint16_t`?

Going forward, would it be OK if I pushed minor fixes (cosmetic nits, mostly) directly to your branch?

Thank you for your work!

-p.

**Comment by FabianHenneke** [comment]
> If possible, I would prefer to merge these commits separately. Doing so would allow modifications to be merged and tested more quickly. 

This makes sense. Would you want me to resubmit as separate PRs our would you split them manually?

> * I like the `legacy_io` approach. I suppose we could rename it `fixed_rpt_size` (for clarity), or derive it from `io.open != fido_hid_open` (for obscurity ;-));

I prefer the 'for clarity' option since it potentially allows custom io functions to take advantage of the support for report length extraction.

> * Perhaps we could define a MIN counterpart of `MAX_CTAP_REPORT_LEN` (`CTAP_INIT_HEADER_LEN + 1`?), and call them `CTAP_{MIN,MAX}_REPORT_LEN` (for consistency with other CTAP_* definitions);

Good idea, this should improve overall consistency.

> * We can probably get rid of the open/close/read/write wrappers by moving `report_{in,out}_len` inside `fido_dev_t` and defining` fido_hid_report_{in,out}_len()` in our HID backends; see [9753f45](https://github.com/Yubico/libfido2/commit/9753f45ae531175b10833ff7d27ab5af5b5d4461) (untested and incomplete code; just a sketch). In this approach, dev.c and hid_*.c are kept separate and do not share types;

You are right, this probably improves separation overall, even though it means that every backend has to define its own context structure. I like your suggestion PR.

> * Would it make sense to define `report_{in,out}_len` in `fido_dev_t` as `uint16_t`?

As in moving it over from 'fido_dev_io_info_t'? Sure!

> Going forward, would it be OK if I pushed minor fixes (cosmetic nits, mostly) directly to your branch?

Go ahead. :-)

**Comment by martelletto** [comment]
> > If possible, I would prefer to merge these commits separately. Doing so would allow modifications to be merged and tested more quickly.
> 
> This makes sense. Would you want me to resubmit as separate PRs our would you split them manually?

I'd prefer PRs, but am fine splitting the commits manually. As you wish.

-p.

**Comment by FabianHenneke** [comment]
> I'd prefer PRs, but am fine splitting the commits manually. As you wish.

I have submitted the first commit, with your suggested changes, as a new PR #172. We could leave this PR open to track the submissions of the individual PRs.

**Comment by FabianHenneke** [comment]
Now that the base PR has been merged it's probably best to close this and deal with the remaining PRs one by one.

**Pull request closed without merging**